### PR TITLE
Feature: Exponential Backoff in Retry Middleware

### PR DESCRIPTION
**Pull request by danieladams456**
### What does this PR do?

This change is a late submission from the Traefik Hackaethon.  (@notsureifkevin) I added exponential backoff to the retry middleware.

### Motivation

#5282 Looked like a fun issue to implement in Traefik was in the plugin bounties [suggested list](https://traefik.io/blog/traefik-hackaethon-2020-middleware-plugins-brain-dump/)! I just started learning Go a month ago, and this is my first "real" project.

### More

- [x] Added/updated tests
  - unit test cases copying your existing main list test cases and running through those all again with a tiny backoff to exercise that code path without slowing things down
  - integration test asserting the elapsed time was what was expected, added a time check to the non-backoff integration test to make sure no regressions there
- [x] Added/updated documentation
  - markdown docs
  - reference docs: it looked like I should hand edit those. `make generate-webui` would remove the line containing `randomizationFactor` in `docs/content/reference/dynamic-configuration/kv-ref.md` that I manually added, but it would not add lines if they were missing?

### Additional Notes

- During the hackathon, it was pointed out that I should use [cenkalti/backoff](https://github.com/cenkalti/backoff) as the backoff implementation. It looks like a really solid package! I saw other uses of that package in the code base use `backoff.RetryNotify()`. I started to refactor `retry.ServeHTTP()` to use that style, but it was going to cause a lot of change to the code flow. Right now it's a while true loop with specific breakout logic in `httptrace.ClientTrace` objects. This way using `ExponentialBackoff` directly and using my own `Sleep`s kept the code flow the same and only entered the `if` statement if backoff was enabled. The refactor would make everything flow through the backoff and just use a `ZeroBackoff` if not enabled. I would be happy to create and self assign a new issue if using `RetryNotify()` is the preferred approach. Then that diff could be reviewed by itself.
- I saw the script tags on the config, and it was pointed out to me at the hackathon that they are used with `make generate-crd`. I was getting errors about copying into a `v2` folder when none existed (`cp -r "${REPO_ROOT}"/"${TRAEFIK_MODULE_VERSION:?}"/* "${REPO_ROOT}";`). I figured I was probably missing something or doing some commands in the wrong order.
- For a long time I was trying to run the retry integration tests via go, and it would show "0 tests ran". Much later I found in some test runner code that only certain tests can be run in go, and others have to be run via docker. The [docs](https://doc.traefik.io/traefik/contributing/building-testing/#method-2-go) seem to indicate either is an option for any integration test.

### CODE REVIEW

**Comment by rtribotte** [comment]
Hello @danieladams456,

Thanks for your time and the work on this PR !

After discussion with the team, we have some feedbacks.

Exposing all the options of the backoff makes Traefik tied to the cenkalti/backoff library actual implementation.
We always prefer to keep the configuration as simple as possible, and open discussion later as soon as specific cases rises.

So, as a first step, we think that configuring the retry middleware only through the max attempts and the backoff initial interval is enough.
This interval should have a default value, and exponential backoff be the new default behavior for the retry middleware.

The multiplier would be computed to be consistent with the number of attempts. The computation could be to wait the double of the initial interval on the last attempt.

The randomization factor should fit most use cases with the default value.

So the configuration would be:
```yml
retry:
   attempts: 42
   initialInterval: 42
```

The way you are using cenkalti/backoff to obtain the interval to sleep seems valid. Further rewriting of the retry middleware is probably better for another step.

**Comment by danieladams456** [comment]
Hey @rtribotte!  I would have tagged you since if I had known your github username. 😄   Thank you for the help during the hackathon and for this pointer.  Tuning it down to that simple configuration sounds great to me.  It should make it nice and easy to use.  It will also vastly simplify my docs and my inability to explain "backoff is activated if any one of these options is set."

Also cool on making it the default behavior!  With it being default, I feel like it should be relatively short to not be a huge change for existing users.  In this case maybe 100-250ms is a good default initial backoff?  Good idea about deriving the multiplier.

If you could get back with my on what you think the initial backoff number would be, I will start working on the changes tonight.

**Comment by danieladams456** [comment]
@rtribotte How would the zero value work on `initialInterval`?  If the attribute is not specified it should be defaulted, but in the middleware construction that would look the same as setting `0` for no backoff.  Would that require having a sentinel value of `-1` for disabling?  That might be non intuitive.  Do you have any ideas on the UX?

**Comment by danieladams456** [comment]
Another idea... What if we said "the max interval is the number of retires multiplied by the initial interval." If you get up to 3-4 retires with max being double, especially a shorter initial interval, it will start to look more linear than exponential. Given that the first retry will solve a large percentage of problems, I would want that to happen as quickly as possible (like 100ms) but 3-4 retires with a max interval of 200ms feels a bit "cramped." What do you all think?

**Comment by rtribotte** [comment]
Hello @danieladams456,

Thanks for your modifications.

Sleeping should happen after a first attempt, so i think you must move this part after the `serveHTTP`.

The default value for the minimal interval could be 100ms. It's very opinionated and should be discuss further, but since it's a default value, this must be configured to fit each use case.

Since it's a duration, I think that the zero value means the deactivation of the exponential backoff, and `NextBackOff()` must return `0s`. IMHO, I think this is easy to understand from the user point of view.

For the multiplier, sorry, i was not sufficiently specific, the double of the initial interval was to illustrate a fixed factor. 
For the rule `maxInterval = maxAttempts * initialInterval`, i think this could be simple enough and less limiting than a fixed factor.
But the fixed factor could be used to better control the velocity to reach the max interval on the last attempts.
Besides, with the rule `maxInterval = maxAttempts * initialInterval`, you cannot really control the velocity.
Maybe a fixed factor relatively high combined with the number of attempts give more control to the user. WDYT ?

**Comment by danieladams456** [comment]
Right now the sleep only triggers on the way "into" the second due to the `attempts >= 2` condition.  I think that was a holdover from when I had my own stateless helper function based on attempt number computing the backoff.  I wanted to do it pre increment since it is already 1-indexed and I was subtracting 1.  The math was harder to follow if I subtracted 2.  Now that I'm using the backoff package and the state is self-contained, I should be able to move it to the very end of the loop.  I will try that.

100ms sounds good to me on a default value.

The question I had about zero value was how to differentiate the first two configs from the third.  The first two would indicate wanting no backoff, but the third should indicate defaulting.  Functionally when reading config to create the middleware, wouldn't all three look the same?

```yaml
retry:
   attempts: 42
   initialInterval: 0
```
or
```yaml
retry:
   attempts: 42
   initialInterval: 0s
```
versus
```yaml
retry:
   attempts: 42
```

I think the rule `maxInterval = maxAttempts * initialInterval` could be a relatively sane default.  Personally I think more in terms of a multiplier than max interval if I was tuning this for my own app.  It makes more sense in my head to compute the max interval from the multiplier than trying to think the other way around. "What is going to be my sequence of values in the ramp up given this max interval and number of attempts?"  Therefore, if we were going to expose one of those two numbers to the user for further configuration, multiplier makes more sense to me.  If there is feedback that more configuration is needed, we could add the logic `if multiplier > 0, then use that, else use default rule`.

**Comment by rtribotte** [comment]
Hello @danieladams456,

Sorry to rewind a bit.
After a second thought and a talk with the team, be breaking by having the exponential backoff by default with an opinionated value for the default for the initial interval (100ms) must be avoided.
We prefer to start with a default value of 0 for the initial interval that would keep the actual behavior as a first step.

So there is no more differences between:
```yaml
retry:
   attempts: 42
   initialInterval: 0s
```
and
```yaml
retry:
   attempts: 42
```

Regarding the multiplier and the value of the max interval, here is what we think should be good at a first step:
- `maxInterval = 2 * initialInterval`
- `multiplier = math.Pow(2, 1/float64(attempts))`

I agree that this looks like linear because of the relatively low value for the maxInterval (2 times the initial interval), but regarding the context of a request, this is already a lot.

You are totally right for the condition `attempts >= 2`, i've just misread that.

For now, let's do not expose more configuration 👍

**Comment by danieladams456** [comment]
I agree having backoff explicit is non breaking, follows principle of least surprise, and makes the config easier as the zero values line up.

For my calculation I was using `b.Multiplier = math.Pow(2, 1/float64(config.Attempts-1))`.  I think the minus 1 makes it line up since `attempts` is 1-indexed instead of 0-indexed.
```
2 attempts == 1 backoff interval == 2^1 multiplier to get to 2x max
3 attempts == 2 backoff intervals == 2^1/2 multiplier == 1.414 to get to 2x max over two backoffs
```

Does it make the code clearer or more redundant to set `maxInterval = 2 * initialInterval` vs giving the reasoning in a comment above the formula?  With the calculated multiplier and number of attempts, they are equivalent.

**Comment by danieladams456** [comment]
I addressed those issues.  Looks like it's working as intended.
```toml
[http.middlewares.retry.retry]
  attempts = 4
  initialInterval = "500ms"
```
```log
DEBU[2020-10-27T21:05:26-04:00] New attempt 2 for request: /                  middlewareName=retry@file middlewareType=Retry
DEBU[2020-10-27T21:05:26-04:00] sleeping 468.857094ms seconds before attempt 2  middlewareType=Retry middlewareName=retry@file
DEBU[2020-10-27T21:05:26-04:00] New attempt 3 for request: /                  middlewareType=Retry middlewareName=retry@file
DEBU[2020-10-27T21:05:26-04:00] sleeping 582.485122ms seconds before attempt 3  middlewareName=retry@file middlewareType=Retry
DEBU[2020-10-27T21:05:27-04:00] New attempt 4 for request: /                  middlewareName=retry@file middlewareType=Retry
DEBU[2020-10-27T21:05:27-04:00] sleeping 941.982095ms seconds before attempt 4  middlewareType=Retry middlewareName=retry@file
```

I stripped out the test generation logic and manually added another one into the test case array with backoff enabled.  How should we proceed with integration testing?  Previously with passing more config items in, I was able to set a randomization factor or 0 and give a small allowed window for the retries to finish.  I don't know that the test would prove that much given the noise from randomness.  About all we could say is "there was some level of artificial waiting between 1.5 and 3 seconds".  Would that type of a range suffice, or do you have any alternative ideas?

**Comment by rtribotte** [comment]
Hello @danieladams456,

Thanks for you changes.
Regarding the implementation, to handle the context cancelation, using a select and `time.After` can be a good solution and be something like :
```go
attempts := 1
backOff := r.newBackOff()
currentInterval := 0 * time.Millisecond
for {
	select {
	case <-time.After(currentInterval):

		shouldRetry := attempts < r.attempts
		retryResponseWriter := newResponseWriter(rw, shouldRetry)

		// Disable retries when the backend already received request data
		trace := &httptrace.ClientTrace{
			WroteHeaders: func() {
				retryResponseWriter.DisableRetries()
			},
			WroteRequest: func(httptrace.WroteRequestInfo) {
				retryResponseWriter.DisableRetries()
			},
		}
		newCtx := httptrace.WithClientTrace(req.Context(), trace)

		r.next.ServeHTTP(retryResponseWriter, req.WithContext(newCtx))

		if !retryResponseWriter.ShouldRetry() {
			return
		}

		currentInterval = backOff.NextBackOff()

		attempts++

		log.FromContext(middlewares.GetLoggerCtx(req.Context(), r.name, typeName)).
			Debugf("New attempt %d for request: %v", attempts, req.URL)

		r.listener.Retried(req, attempts)

	case <-req.Context().Done():
		return
	}
}
```

Regarding the unit tests, maybe adding a unit test to check the backOff configuration is enough.
For the integration tests, let me take another look later to give you more feedbacks, but maybe ensuring that the duration is greater or equal than the number of attempts multiply by the initial interval is enough, other assertions are maybe two risky and will potentially raise the chances to have a flaky test.

There is also a minor change to make to the `backoff.toml` fixture configuration file which still uses old options that are no more exposed.

**Comment by danieladams456** [comment]
Thank you for the elegant wait+cancellation solution @rtribotte!  At first I was wondering if potentially slow connection attempt time spent in `ServerHTTP()` would count against the effective timer, but I see that time starts as it enters the `select` statement after the previous `for` iteration has completed.

I fixed the integration test and docs.  I will hit the additional unit test later tonight, and then should be ready to proceed with the next phase. 😄

**Comment by danieladams456** [comment]
I think we are ready for review.  I added some more tests with backoff to the main test list.  Previously I had generated those from looping through and appending.  Which style would you prefer?

Was that what you were referring to with your unit test comment or something like this?
```go
func TestBackoffFromConfig(t *testing.T) {
	initialInterval := time.Millisecond * 50
	retryH, err := New(context.Background(), nil, dynamic.Retry{Attempts: 3, InitialInterval: ptypes.Duration(initialInterval)}, nil, "traefikTest")
	require.NoError(t, err)

	castFail := retry(retryH)
	b := newBackOff()
	assert.Equal(b.NextBackoff(), initialInterval)
	assert.Greater(b.NextBackoff(), initialInterval)
}
```
I can't do that cast even though it "really" is the type since the function only returns a `http.Handler`.  Also is that too brittle of a test or not?

Let me know if the overall testing approach looks good to you or if there is anything missing. Thanks!

**Comment by danieladams456** [comment]
The previous CI run looked pretty good before the lint rule stopped it... Not sure if all these failures are a fluke?

**Comment by rtribotte** [comment]
Hello @danieladams456,

Thanks for the changes.
I prefer the explicit way to declare the test cases, even though it's more verbose.

For the test you submitted, i'm sorry i missed the point of the cast, what is your intention here ?

The ci failed because of `label_test.go` which need to be fixed since the ìnitialInterval` has a default value.

**Comment by danieladams456** [comment]
Sure, I will revert the test structure.  Would you like the equivalent backoff test for all 5 test cases or just 1-2.  I'm thinking might as well go for all?

The cast would let me pull the backoff object out of the middleware to assert about the timings.  But the `New` function returns just a `http.Handler` so all the other properties are encapsulated.  I'm thinking that is not really necessary anyway as the integration tests cover timing.  It's kind of trying to force test private methods instead of the public API so probably not a good pattern?

I will look at that labels test.

**Comment by rtribotte** [comment]
I have no strong opinion, because i need more time to affirm that all the tests have to be duplicate, but i guess that focusing on what matter to be tested is the rule ;-)
I also think that this test is not necessary 👍

**Comment by danieladams456** [comment]
Tests are all passing now.

**Review by kevinpollet — CHANGES_REQUESTED** [review]
Nice work!

I've made some suggestions.

File: pkg/middlewares/retry/retry_test.go [hunk 1]
```diff
@@ -64,6 +68,16 @@ func TestRetry(t *testing.T) {
 			amountFaultyEndpoints: 3,
 		},
 	}
+	for _, v := range testCases {
+		withBackoff := testCase{
```
[the remark this answers is not in the document]
Comment (danieladams456) [reply pkg/middlewares/retry/retry_test.go:72]: Sounds good.  I'll let you all decide if you think all the combinations should be tried with backoff to exercise that code path, or just 1-2.  The above comment can also be resolved about using an inline type definition now that I'm not constructing test items dynamically.

File: pkg/middlewares/retry/retry.go [hunk 2]
```diff
@@ -50,11 +60,30 @@ func New(ctx context.Context, next http.Handler, config dynamic.Retry, listener
 		return nil, fmt.Errorf("incorrect (or empty) value for attempt (%d)", config.Attempts)
 	}
 
+	// a new backoff instance should be created for every request
+	newBackOff := func() backoff {
+		if config.Attempts < 2 || config.InitialInterval == 0 {
+			b := new(bo.ZeroBackOff)
+			return b
```
[the remark this answers is not in the document]
Comment (danieladams456) [reply pkg/middlewares/retry/retry.go:67]: Much cleaner

File: pkg/middlewares/retry/retry.go [hunk 3]
```diff
@@ -50,11 +60,30 @@ func New(ctx context.Context, next http.Handler, config dynamic.Retry, listener
 		return nil, fmt.Errorf("incorrect (or empty) value for attempt (%d)", config.Attempts)
 	}
 
+	// a new backoff instance should be created for every request
+	newBackOff := func() backoff {
```
[the remark this answers is not in the document]
Comment (danieladams456) [reply pkg/middlewares/retry/retry.go:64]: I think it was a holdover where I was taking in 4 config parameters and not wanting to cause as much property bloat.  Now it's just 1 property.  Having a struct function is probably better.  I'll update to that.

**Review by kevinpollet — APPROVED** [review]
LGTM

File: pkg/middlewares/retry/retry.go [hunk 4]
```diff
@@ -72,34 +82,61 @@ func (r *retry) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
...
+
+func (r *retry) newBackOff() nexter {
+	if r.attempts < 2 || r.initialInterval <= 0 {
+		return &backoff.ZeroBackOff{}
 	}
+
+	b := backoff.NewExponentialBackOff()
+	b.InitialInterval = r.initialInterval
+
+	// calculate the multiplier that will set MaxInterval = 2*InitialInterval
```
Comment (rtribotte) [inline pkg/middlewares/retry/retry.go:134]: Thanks for your previous changes, i think the comment should clarify the purpose here, and not confuse with the max interval, what do you think ?
```suggestion
	// calculate the multiplier for the given number of attempts
	// so that applying the multiplier for the given number of attempts will not exceed 2 times the initial interval
	// it allows to control the progression along the attempts
```

**Comment by danieladams456** [comment]
I don't know if we want to change this PR so late in the process... I had previously deferred what I had guessed would be a larger refactor to `backoff.RetryNotify`.  Originally it was just dropping in a sleep if statement, but we needed to implement context cancellation.  That added in the select statement to do the wait and cancel.

I have a refactor ready to use `backoff.RetryNotify` which I think [makes the flow clearer](https://github.com/danieladams456/traefik/blob/992692e9d4c7fa2e61219b2e9cc20b534b7c5b11/pkg/middlewares/retry/retry.go#L77).  The existing unit/integration tests are passing.  I tested cancellation manually.  Is that one we need to add to the integration test suite?  I can push this code to this PR branch, or can immediately submit a followup PR.  Let me know which one you'd like me to do.

**Comment by rtribotte** [comment]
Hello @danieladams456,

Thanks for your work and your reactivity all along the work on this PR.
I'll discuss that with the team, but my guess for now is that we should let this PR as is, since we are reaching the state of acceptance and it already took quite some efforts !
Even if it's probably best to make the "flow" clearer, we would have to go through a new cycle of review, especially looking for unforeseen side effects.
Nonetheless, you definitely can submit your work in another PR, if you feel that it worth to do it, it is always welcome! That being said, you'll understand that I cannot make the commitment that it will be accepted.

**Review by rtribotte — APPROVED** [review]
LGTM 👍

File: pkg/config/label/label_test.go [hunk 5]
```diff
@@ -121,6 +121,7 @@ func TestDecodeConfiguration(t *testing.T) {
 		"traefik.http.middlewares.Middleware15.replacepathregex.regex":                             "foobar",
 		"traefik.http.middlewares.Middleware15.replacepathregex.replacement":                       "foobar",
 		"traefik.http.middlewares.Middleware16.retry.attempts":                                     "42",
+		"traefik.http.middlewares.Middleware16.retry.initialInterval":                              "1s",
```
Comment (jbdoumenjou) [inline pkg/config/label/label_test.go:124]:
```suggestion
		"traefik.http.middlewares.Middleware16.retry.initialinterval":                              "1s",
```

**Review by jbdoumenjou — CHANGES_REQUESTED** [review]
Thanks :+1: Some minor comments to be homogeneous.

**Review by jbdoumenjou — APPROVED** [review]
LGTM :+1:
